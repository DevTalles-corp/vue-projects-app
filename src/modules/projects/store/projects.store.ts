import { defineStore } from 'pinia';
import { computed, ref } from 'vue';
import type { Project } from '../interfaces/project.interface';
import { v4 as uuidv4 } from 'uuid';
import { useLocalStorage } from '@vueuse/core';

export const useProjectsStore = defineStore('projects', () => {
  const projects = ref(useLocalStorage<Project[]>('projects', []));

  const addProject = (name: string) => {
    if (name.length === 0) return;

    projects.value.push({
      id: uuidv4(),
      name: name,
      tasks: [],
    });
  };

  const addTaskToProject = (projectId: string, taskName: string) => {
    if (taskName.trim().length === 0) return;

    const project = projects.value.find((p) => p.id === projectId);
    if (!project) return;

    project.tasks.push({
      id: uuidv4(),
      name: taskName,
    });
  };

  const toggleTask = (projectId: string, taskId: string) => {
    const project = projects.value.find((p) => p.id === projectId);
    if (!project) return;

    const task = project.tasks.find((t) => t.id === taskId);
    if (!task) return;

    task.completedAt = task.completedAt ? undefined : new Date();
  };

  return {
    // Properties
    projects,

    // Getters
    projectList: computed(() => [...projects.value]),
    noProjects: computed(() => projects.value.length === 0),

    projectsWithCompletion: computed(() => {
      return projects.value.map((project) => {
        const total = project.tasks.length;
        const completed = project.tasks.filter((t) => t.completedAt).length;
        const completion = total === 0 ? 0 : (completed / total) * 100;

        return {
          id: project.id,
          name: project.name,
          taskCount: total,
          completion: Math.round(completion),
        };
      });
      // {
      //   id,
      //   name,
      //   taskCount,
      //   completion,
      // }
    }),

    // Actions
    addProject,

    addTaskToProject,
    toggleTask,
  };
});
