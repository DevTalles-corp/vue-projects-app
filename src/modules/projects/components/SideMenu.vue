<template>
  <aside class="bg-base-200 w-72 min-h-screen">
    <h2 class="text-lg font-bold mx-4">
      <RouterLink to="/"> Proyectos </RouterLink>
    </h2>
    <p v-if="projectsStore.noProjects" class="text-sm text-gray-500 mx-4">No hay proyectos</p>

    <!-- Menu -->
    <ul v-else class="menu">
      <li v-for="project in projectsStore.projectList" :key="project.id">
        <template v-if="project.tasks.length > 0">
          <details>
            <summary>
              <RouterLink :to="`/project/${project.id}`">
                {{ project.name }}
              </RouterLink>
            </summary>
            <ul>
              <li v-for="task in project.tasks" :key="task.id">
                <RouterLink :to="`/project/${project.id}`">{{ task.name }}</RouterLink>
              </li>
            </ul>
          </details>
        </template>

        <template v-else>
          <RouterLink :to="`/project/${project.id}`">
            {{ project.name }}
          </RouterLink>
        </template>
      </li>
    </ul>
  </aside>
</template>

<script lang="ts" setup>
import { useProjectsStore } from '../store/projects.store';

const projectsStore = useProjectsStore();
</script>
