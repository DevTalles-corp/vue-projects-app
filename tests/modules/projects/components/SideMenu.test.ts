import { nextTick } from 'vue';
import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import SideMenu from '@/modules/projects/components/SideMenu.vue';
import { useProjectsStore } from '@/modules/projects/store/projects.store';
import { fakeProjects } from '../../../mocks/projects.fake';

describe('<SideMenu />', () => {
  const wrapper = mount(SideMenu, {
    global: {
      stubs: ['RouterLink'],
      plugins: [createTestingPinia()],
    },
  });

  const store = useProjectsStore();

  beforeEach(() => {
    store.$patch({
      projects: [],
    });
  });

  test('should render with no projects', () => {
    expect(wrapper.html()).toContain('No hay proyectos');
  });

  test('should render with no projects', async () => {
    store.$patch({
      projects: fakeProjects,
    });

    await nextTick();

    expect(wrapper.html()).not.toContain('No hay proyectos');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
