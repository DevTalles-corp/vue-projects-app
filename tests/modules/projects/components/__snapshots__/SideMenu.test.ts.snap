// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SideMenu /> > should render with no projects 1`] = `
"<aside class="bg-base-200 w-72 min-h-screen">
  <h2 class="text-lg font-bold mx-4">
    <router-link-stub to="/"></router-link-stub>
  </h2><!-- Menu -->
  <ul class="menu">
    <li>
      <details>
        <summary>
          <router-link-stub to="/project/1"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/1"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
    <li>
      <router-link-stub to="/project/2"></router-link-stub>
    </li>
    <li>
      <details>
        <summary>
          <router-link-stub to="/project/3"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/3"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/3"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
    <li>
      <details>
        <summary>
          <router-link-stub to="/project/4"></router-link-stub>
        </summary>
        <ul>
          <li>
            <router-link-stub to="/project/4"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/4"></router-link-stub>
          </li>
          <li>
            <router-link-stub to="/project/4"></router-link-stub>
          </li>
        </ul>
      </details>
    </li>
  </ul>
</aside>"
`;
